src/charm.py

```diff
@@ -699,7 +719,12 @@ def _process_ingresses(self, ingresses):
                 # Ensure that the exact same route for this route was not yet defined.
                 defined_paths = ingress_paths[rule.host]
                 for path in rule.http.paths:
-                    duplicate_paths = list(filter(lambda p: p.path == path.path, defined_paths))
+                    duplicate_paths = list(
```

Comment:
This function looks like it would benefit from refactoring, there are a lot of for loops in for loops, if statement etc